[WIP]: Change prio for rlbox sandboxing

Hyperion101010 opened the pull request:
This patch adds support to use nss functions
  in rlbox wasm sandbox. We can thus enable the
  use of Prio as a sandboxed lib for Firefox.

Commit:
[WIP]: Change prio for rlbox sandboxing

  This patch adds support to use nss functions
  in rlbox wasm sandbox. We can thus enable the
  use of Prio as a sandboxed lib for Firefox.

ekr commented:
💯

rhelmer commented:
This builds OK, but is failing running tests:
https://travis-ci.org/github/mozilla/libprio/jobs/679192258#L3819

```
$ ./build/ptest/ptest -v

build/prio/rand.c:31:36: runtime error: member access within misaligned address 0x000000000002 for type 'nss_struct' (aka 'struct nss_data'), which requires 8 byte alignment
0x000000000002: note: pointer points here
<memory cannot be printed>
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior build/prio/rand.c:31:36 in 
/Users/travis/.travis/functions: line 109: 52920 Abort trap: 6           ./build/ptest/ptest -v
The command "./build/ptest/ptest -v" exited with 134.
```

@henrycg might be interested ^

henrycg commented:
It'd be great to have libprio sandboxed in FF. Let me know when you'd like me to review this.

Hyperion101010 commented:
Did some changes in the weekend. First was hitting some errors but later found out I forgot to assign the functions and also add it up in the build script. 
Fixed on my local and got success on the tests.
@shravanrn  @rhelmer  please check.

Hyperion101010 commented:
Doesn't amend run a test?

Commit:
[WIP]: Change prio for rlbox sandboxing

 Follow up.

rhelmer commented:
This passes tests now, the Linux build is broken for a different reason. I'll fix that rn in PR #86

Hyperion101010 commented:
Ok

rhelmer committed:
upgrade Travis CI distro to Ubuntu Bionic (#86)

rhelmer commented:
@Hyperion101010 fixed in PR #86, please rebase master and push this branch. Sorry for the hassle.

shravanrn commented on include/callbk.h:
```diff
@@ -0,0 +1,36 @@
+
+#include <nss.h>
+
+typedef struct nss_data nss_struct;
+
+struct nss_data
+{
+    /* data */
+    NSSInitContext* (*nssinit)(const char *configdir,
+                                       const char *certPrefix, const char *keyPrefix,
+                                       const char *secmodName, NSSInitParameters *initParams, PRUint32 flags);
+
+    PRBool (*nssisinit)();
+
+    SECStatus (*nssshutdown)(NSSInitContext *context);
+};
```
```suggestion
typedef struct
{
    /* data */
    NSSInitContext* (*nssinit)(const char *configdir,
                                       const char *certPrefix, const char *keyPrefix,
                                       const char *secmodName, NSSInitParameters *initParams, PRUint32 flags);

    PRBool (*nssisinit)();

    SECStatus (*nssshutdown)(NSSInitContext *context);
} nss_struct;
```

shravanrn commented on prio/rand.c:
```diff
@@ -16,18 +16,22 @@
 
 SECStatus
-rand_init(void)
+rand_init(nss_struct *nss_pointer)
```
This approach assumes that rand_init is not invoked in parallel. Did you confirm if this is the case, or if the API allows parallel calls to rand_init? If it doesn't this is fine, if it does, you need a lock somewhere.

Hyperion101010 replied:
I looked in prio(pclient) and tests, there are no parallel calls.

shravanrn commented on prio/config.c:
```diff
@@ -95,13 +99,14 @@ PrioConfig_numDataFields(const_PrioConfig cfg)
 Prio_init(void)
 {
-  return rand_init();
+  nss_callbk_init(&nss_pointer);
```
Prio_init needs to allow the caller to specify the nss_struct, otherwise this is not configurable

Hyperion101010 replied:
the tests are written such that they don't allow the user to configure the input. If i were to update it, then it requires some changes in tests and also the pclient itself. It is a random number generator so i think adding configuration won't change anything in the end?

shravanrn replied:
I think you are missing the point here. The purpose of this change is to allow prio to be independent of nss. This way prio can be compiled without nss. A consumer of prio such as firefox would then provide the nss functions via pointers passed in when calling Prio_init. Right now you have hard coded the pointers to be to nss inside the call to `nss_callbk_init` which means prio must once again be compiled with nss. `Prio_init` needs th have the signature `Prio_init(nss_struct *nss_pointer)`

Hyperion101010 replied:
ok then I will change it, so now updating the tests and other depending files.

Commit:
[WIP]: Change prio for rlbox sandboxing #84

 Made changes in nss struct in include/callbk.h

Hyperion101010 commented:
Some commit conflicts on my end.

rhelmer commented:
The current Travis CI failures are just the clang-format check:
https://travis-ci.org/github/mozilla/libprio/jobs/680041898

You can run `./scripts/clang-format.sh` locally. Everything else looks good as far as automated testing.

Commit:
[WIP]: Change prio for rlbox sandboxing #84

   Made changes in the prio lib to pass nss
   functions as pointers. Needed to make changes
   in pclient to not break tests.
   They will be later implemented in wrapper
   for firefox.

Hyperion101010 commented:
So i tried to run clang-format in scripts but it was throwing me errors. 
log ->  *** ERROR:  clang-format 3.9 required

I do have clang-format installed 
hyperion@hyperion-rog:~/intern/gsoc2020/wasm_sandbox/libprio$ clang-format --version
clang-format version 9.0.0-2 (tags/RELEASE_900/final)

So used git-clang-format instead.

Hyperion101010 commented:
Looks like ci uses only ./script/clang-format.sh and any other method might result in incorrect formatting.

Hyperion101010 commented:
Usually running git-clang-format and then uploading fixed my formatting, i tried to do some workaround and run script in ./script/clang-format.sh
it works and throws me some output. 
But it doesn't make any changes inside those files which is expected. 
@rhelmer i might be doing it wrong will you please guide. 

the op of running script
./scripts/clang-format.sh .
Checking files:
include/mprio.h
pclient/main.c
prio/callbk.c
prio/config.c
ptest/mutest.c

*** The following differences were found between the code to commit and the clang-format rules:

--- a/include/mprio.h	2020-04-29 16:10:45.081830336 +0530
+++ b/include/mprio.h	2020-04-29 16:34:38.960857940 +0530
@@ -10,7 +10,8 @@
 #define __PRIO_H__
 
 #ifdef __cplusplus
-extern "C" {
+extern "C"
+{
 #endif
 
 #include <blapit.h>
@@ -24,7 +25,7 @@
 
 /* Seed for a pseudo-random generator (PRG). */
 #define PRG_SEED_LENGTH AES_128_KEY_LENGTH
-typedef unsigned char PrioPRGSeed[PRG_SEED_LENGTH];
+  typedef unsigned char PrioPRGSeed[PRG_SEED_LENGTH];
 
 /* Length of a raw curve25519 public key, in bytes. */
 #define CURVE25519_KEY_LEN 32
@@ -32,272 +33,296 @@
 /* Length of a hex-encoded curve25519 public key, in bytes. */
 #define CURVE25519_KEY_LEN_HEX 64
 
-/*
- * Type for each of the two servers.
- */
-typedef enum { PRIO_SERVER_A, PRIO_SERVER_B } PrioServerId;
-
-/*
- * Opaque types
- */
-typedef struct prio_config* PrioConfig;
-typedef const struct prio_config* const_PrioConfig;
-
-typedef struct prio_server* PrioServer;
-typedef const struct prio_server* const_PrioServer;
-
-typedef struct prio_verifier* PrioVerifier;
-typedef const struct prio_verifier* const_PrioVerifier;
-
-typedef struct prio_packet_verify1* PrioPacketVerify1;
-typedef const struct prio_packet_verify1* const_PrioPacketVerify1;
-
-typedef struct prio_packet_verify2* PrioPacketVerify2;
-typedef const struct prio_packet_verify2* const_PrioPacketVerify2;
-
-typedef struct prio_total_share* PrioTotalShare;
-typedef const struct prio_total_share* const_PrioTotalShare;
-
-typedef SECKEYPublicKey* PublicKey;
-typedef const SECKEYPublicKey* const_PublicKey;
-
-typedef SECKEYPrivateKey* PrivateKey;
-typedef const SECKEYPrivateKey* const_PrivateKey;
-
-/*
- * Initialize and clear random number generator state.
- * You must call Prio_init() before using the library.
- * To avoid memory leaks, call Prio_clear() afterwards.
- */
-SECStatus Prio_init();
-void Prio_clear();
-
-/*
- * PrioConfig holds the system parameters. The two relevant things determined
- * by the config object are:
- *    (1) the number of data fields we are collecting, and
- *    (2) the modulus we use for modular arithmetic.
- * The default configuration uses an 87-bit modulus.
- *
- * The value `nFields` must be in the range `0 < nFields <= max`, where `max`
- * is the value returned by the function `PrioConfig_maxDataFields()` below.
- *
- * The `batch_id` field specifies which "batch" of aggregate statistics we are
- * computing. For example, if the aggregate statistics are computed every 24
- * hours, the `batch_id` might be set to an encoding of the date. The clients
- * and servers must all use the same `batch_id` for each run of the protocol.
- * Each set of aggregate statistics should use a different `batch_id`.
- *
- * `PrioConfig_new` does not keep a pointer to the `batch_id` string that the
- * caller passes in, so you may free the `batch_id` string as soon as
- * `PrioConfig_new` returns.
- */
-PrioConfig PrioConfig_new(int nFields, PublicKey serverA, PublicKey serverB,
-                          const unsigned char* batchId,
-                          unsigned int batchIdLen);
-void PrioConfig_clear(PrioConfig cfg);
-int PrioConfig_numDataFields(const_PrioConfig cfg);
-
-/*
- * Return the maximum number of data fields that the implementation supports.
- */
-int PrioConfig_maxDataFields(void);
-
-/*
- * Create a PrioConfig object with no encryption keys.  This routine is
- * useful for testing, but PrioClient_encode() will always fail when used with
- * this config.
- */
-PrioConfig PrioConfig_newTest(int nFields);
-
-/*
- * We use the PublicKey and PrivateKey objects for public-key encryption. Each
- * Prio server has an associated public key, and the clients use these keys to
- * encrypt messages to the servers.
- */
-SECStatus Keypair_new(PrivateKey* pvtkey, PublicKey* pubkey);
-
-/*
- * Import a new curve25519 public/private key from the raw bytes given.  When
- * importing a private key, you must pass in the corresponding public key as
- * well. The byte arrays given as input should be of length
- * `CURVE25519_KEY_LEN`.
- *
- * These functions will allocate a new `PublicKey`/`PrivateKey` object, which
- * the caller must free using `PublicKey_clear`/`PrivateKey_clear`.
- */
-SECStatus PublicKey_import(PublicKey* pk, const unsigned char* data,
-                           unsigned int dataLen);
-SECStatus PrivateKey_import(PrivateKey* sk, const unsigned char* privData,
-                            unsigned int privDataLen,
-                            const unsigned char* pubData,
-                            unsigned int pubDataLen);
-
-/*
- * Import a new curve25519 public/private key from a hex string that contains
- * only the characters 0-9a-fA-F.
- *
- * The hex strings passed in must each be of length `CURVE25519_KEY_LEN_HEX`.
- * These functions will allocate a new `PublicKey`/`PrivateKey` object, which
- * the caller must free using `PublicKey_clear`/`PrivateKey_clear`.
- */
-SECStatus PublicKey_import_hex(PublicKey* pk, const unsigned char* hexData,
-                               unsigned int dataLen);
-SECStatus PrivateKey_import_hex(PrivateKey* sk,
-                                const unsigned char* privHexData,
-                                unsigned int privDataLen,
-                                const unsigned char* pubHexData,
-                                unsigned int pubDataLen);
-
-/*
- * Export a curve25519 key as a raw byte-array.
- *
- * The output buffer `data` must have length exactly `CURVE25519_KEY_LEN`.
- */
-SECStatus PublicKey_export(const_PublicKey pk, unsigned char* data,
-                           unsigned int dataLen);
-SECStatus PrivateKey_export(PrivateKey sk, unsigned char* data,
-                            unsigned int dataLen);
-
-/*
- * Export a curve25519 key as a NULL-terminated hex string.
- *
- * The output buffer `data` must have length exactly `CURVE25519_KEY_LEN_HEX +
- * 1`.
- */
-SECStatus PublicKey_export_hex(const_PublicKey pk, unsigned char* data,
-                               unsigned int dataLen);
-SECStatus PrivateKey_export_hex(PrivateKey sk, unsigned char* data,
-                                unsigned int dataLen);
-
-void PublicKey_clear(PublicKey pubkey);
-void PrivateKey_clear(PrivateKey pvtkey);
-
-/*
- *  PrioPacketClient_encode
- *
- * Takes as input a pointer to an array (`data_in`) of boolean values
- * whose length is equal to the number of data fields specified in
- * the config. It then encodes the data for servers A and B into a
- * string.
- *
- * NOTE: The caller must free() the strings `for_server_a` and
- * `for_server_b` to avoid memory leaks.
- */
-SECStatus PrioClient_encode(const_PrioConfig cfg, const bool* data_in,
-                            unsigned char** forServerA, unsigned int* aLen,
-                            unsigned char** forServerB, unsigned int* bLen);
-
-/*
- * Generate a new PRG seed using the NSS global randomness source.
- * Use this routine to initialize the secret that the two Prio servers
- * share.
- */
-SECStatus PrioPRGSeed_randomize(PrioPRGSeed* seed);
-
-/*
- * The PrioServer object holds the state of the Prio servers.
- * Pass in the _same_ secret PRGSeed when initializing the two servers.
- * The PRGSeed must remain secret to the two servers.
- */
-PrioServer PrioServer_new(const_PrioConfig cfg, PrioServerId serverIdx,
-                          PrivateKey serverPriv,
-                          const PrioPRGSeed serverSharedSecret);
-void PrioServer_clear(PrioServer s);
-
-/*
- * After receiving a client packet, each of the servers generate
- * a PrioVerifier object that they use to check whether the client's
- * encoded packet is well formed.
- */
-PrioVerifier PrioVerifier_new(PrioServer s);
-void PrioVerifier_clear(PrioVerifier v);
-
-/*
- * Read in encrypted data from the client, decrypt it, and prepare to check the
- * request for validity.
- */
-SECStatus PrioVerifier_set_data(PrioVerifier v, unsigned char* data,
-                                unsigned int dataLen);
-
-/*
- * Generate the first packet that servers need to exchange to verify the
- * client's submission. This should be sent over a TLS connection between the
- * servers.
- */
-PrioPacketVerify1 PrioPacketVerify1_new(void);
-void PrioPacketVerify1_clear(PrioPacketVerify1 p1);
-
-SECStatus PrioPacketVerify1_set_data(PrioPacketVerify1 p1,
-                                     const_PrioVerifier v);
-
-SECStatus PrioPacketVerify1_write(const_PrioPacketVerify1 p,
-                                  msgpack_packer* pk);
-SECStatus PrioPacketVerify1_read(PrioPacketVerify1 p, msgpack_unpacker* upk,
-                                 const_PrioConfig cfg);
-
-/*
- * Generate the second packet that the servers need to exchange to verify the
- * client's submission. The routine takes as input the PrioPacketVerify1
- * packets from both server A and server B.
- *
- * This should be sent over a TLS connection between the servers.
- */
-PrioPacketVerify2 PrioPacketVerify2_new(void);
-void PrioPacketVerify2_clear(PrioPacketVerify2 p);
-
-SECStatus PrioPacketVerify2_set_data(PrioPacketVerify2 p2, const_PrioVerifier v,
-                                     const_PrioPacketVerify1 p1A,
-                                     const_PrioPacketVerify1 p1B);
-
-SECStatus PrioPacketVerify2_write(const_PrioPacketVerify2 p,
-                                  msgpack_packer* pk);
-SECStatus PrioPacketVerify2_read(PrioPacketVerify2 p, msgpack_unpacker* upk,
-                                 const_PrioConfig cfg);
-
-/*
- * Use the PrioPacketVerify2s from both servers to check whether
- * the client's submission is well formed.
- */
-SECStatus PrioVerifier_isValid(const_PrioVerifier v, const_PrioPacketVerify2 pA,
-                               const_PrioPacketVerify2 pB);
-
-/*
- * Each of the two servers calls this routine to aggregate the data
- * submission from a client that is included in the PrioVerifier object.
- *
- * IMPORTANT: This routine does *not* check the validity of the client's
- * data packet. The servers must execute the verification checks
- * above before aggregating any client data.
- */
-SECStatus PrioServer_aggregate(PrioServer s, PrioVerifier v);
-
-/*
- * After the servers have aggregated data packets from "enough" clients
- * (this determines the anonymity set size), each server runs this routine
- * to get a share of the aggregate statistics.
- */
-PrioTotalShare PrioTotalShare_new(void);
-void PrioTotalShare_clear(PrioTotalShare t);
-
-SECStatus PrioTotalShare_set_data(PrioTotalShare t, const_PrioServer s);
-
-SECStatus PrioTotalShare_write(const_PrioTotalShare t, msgpack_packer* pk);
-SECStatus PrioTotalShare_read(PrioTotalShare t, msgpack_unpacker* upk,
-                              const_PrioConfig cfg);
-
-/*
- * Read the output data into an array of unsigned longs. You should
- * be sure that each data value can fit into a single `unsigned long`
- * and that the pointer `output` points to a buffer large enough to
- * store one long per data field.
- *
- * This function returns failure if some final data value is too
- * long to fit in an `unsigned long`.
- */
-SECStatus PrioTotalShare_final(const_PrioConfig cfg, unsigned long long* output,
-                               const_PrioTotalShare tA,
-                               const_PrioTotalShare tB);
+  /*
+   * Type for each of the two servers.
+   */
+  typedef enum
+  {
+    PRIO_SERVER_A,
+    PRIO_SERVER_B
+  } PrioServerId;
+
+  /*
+   * Opaque types
+   */
+  typedef struct prio_config* PrioConfig;
+  typedef const struct prio_config* const_PrioConfig;
+
+  typedef struct prio_server* PrioServer;
+  typedef const struct prio_server* const_PrioServer;
+
+  typedef struct prio_verifier* PrioVerifier;
+  typedef const struct prio_verifier* const_PrioVerifier;
+
+  typedef struct prio_packet_verify1* PrioPacketVerify1;
+  typedef const struct prio_packet_verify1* const_PrioPacketVerify1;
+
+  typedef struct prio_packet_verify2* PrioPacketVerify2;
+  typedef const struct prio_packet_verify2* const_PrioPacketVerify2;
+
+  typedef struct prio_total_share* PrioTotalShare;
+  typedef const struct prio_total_share* const_PrioTotalShare;
+
+  typedef SECKEYPublicKey* PublicKey;
+  typedef const SECKEYPublicKey* const_PublicKey;
+
+  typedef SECKEYPrivateKey* PrivateKey;
+  typedef const SECKEYPrivateKey* const_PrivateKey;
+
+  /*
+   * Initialize and clear random number generator state.
+   * You must call Prio_init() before using the library.
+   * To avoid memory leaks, call Prio_clear() afterwards.
+   */
+  SECStatus Prio_init();
+  void Prio_clear();
+
+  /*
+   * PrioConfig holds the system parameters. The two relevant things determined
+   * by the config object are:
+   *    (1) the number of data fields we are collecting, and
+   *    (2) the modulus we use for modular arithmetic.
+   * The default configuration uses an 87-bit modulus.
+   *
+   * The value `nFields` must be in the range `0 < nFields <= max`, where `max`
+   * is the value returned by the function `PrioConfig_maxDataFields()` below.
+   *
+   * The `batch_id` field specifies which "batch" of aggregate statistics we are
+   * computing. For example, if the aggregate statistics are computed every 24
+   * hours, the `batch_id` might be set to an encoding of the date. The clients
+   * and servers must all use the same `batch_id` for each run of the protocol.
+   * Each set of aggregate statistics should use a different `batch_id`.
+   *
+   * `PrioConfig_new` does not keep a pointer to the `batch_id` string that the
+   * caller passes in, so you may free the `batch_id` string as soon as
+   * `PrioConfig_new` returns.
+   */
+  PrioConfig PrioConfig_new(int nFields,
+                            PublicKey serverA,
+                            PublicKey serverB,
+                            const unsigned char* batchId,
+                            unsigned int batchIdLen);
+  void PrioConfig_clear(PrioConfig cfg);
+  int PrioConfig_numDataFields(const_PrioConfig cfg);
+
+  /*
+   * Return the maximum number of data fields that the implementation supports.
+   */
+  int PrioConfig_maxDataFields(void);
+
+  /*
+   * Create a PrioConfig object with no encryption keys.  This routine is
+   * useful for testing, but PrioClient_encode() will always fail when used with
+   * this config.
+   */
+  PrioConfig PrioConfig_newTest(int nFields);
+
+  /*
+   * We use the PublicKey and PrivateKey objects for public-key encryption. Each
+   * Prio server has an associated public key, and the clients use these keys to
+   * encrypt messages to the servers.
+   */
+  SECStatus Keypair_new(PrivateKey* pvtkey, PublicKey* pubkey);
+
+  /*
+   * Import a new curve25519 public/private key from the raw bytes given.  When
+   * importing a private key, you must pass in the corresponding public key as
+   * well. The byte arrays given as input should be of length
+   * `CURVE25519_KEY_LEN`.
+   *
+   * These functions will allocate a new `PublicKey`/`PrivateKey` object, which
+   * the caller must free using `PublicKey_clear`/`PrivateKey_clear`.
+   */
+  SECStatus PublicKey_import(PublicKey* pk,
+                             const unsigned char* data,
+                             unsigned int dataLen);
+  SECStatus PrivateKey_import(PrivateKey* sk,
+                              const unsigned char* privData,
+                              unsigned int privDataLen,
+                              const unsigned char* pubData,
+                              unsigned int pubDataLen);
+
+  /*
+   * Import a new curve25519 public/private key from a hex string that contains
+   * only the characters 0-9a-fA-F.
+   *
+   * The hex strings passed in must each be of length `CURVE25519_KEY_LEN_HEX`.
+   * These functions will allocate a new `PublicKey`/`PrivateKey` object, which
+   * the caller must free using `PublicKey_clear`/`PrivateKey_clear`.
+   */
+  SECStatus PublicKey_import_hex(PublicKey* pk,
+                                 const unsigned char* hexData,
+                                 unsigned int dataLen);
+  SECStatus PrivateKey_import_hex(PrivateKey* sk,
+                                  const unsigned char* privHexData,
+                                  unsigned int privDataLen,
+                                  const unsigned char* pubHexData,
+                                  unsigned int pubDataLen);
+
+  /*
+   * Export a curve25519 key as a raw byte-array.
+   *
+   * The output buffer `data` must have length exactly `CURVE25519_KEY_LEN`.
+   */
+  SECStatus PublicKey_export(const_PublicKey pk,
+                             unsigned char* data,
+                             unsigned int dataLen);
+  SECStatus PrivateKey_export(PrivateKey sk,
+                              unsigned char* data,
+                              unsigned int dataLen);
+
+  /*
+   * Export a curve25519 key as a NULL-terminated hex string.
+   *
+   * The output buffer `data` must have length exactly `CURVE25519_KEY_LEN_HEX +
+   * 1`.
+   */
+  SECStatus PublicKey_export_hex(const_PublicKey pk,
+                                 unsigned char* data,
+                                 unsigned int dataLen);
+  SECStatus PrivateKey_export_hex(PrivateKey sk,
+                                  unsigned char* data,
+                                  unsigned int dataLen);
+
+  void PublicKey_clear(PublicKey pubkey);
+  void PrivateKey_clear(PrivateKey pvtkey);
+
+  /*
+   *  PrioPacketClient_encode
+   *
+   * Takes as input a pointer to an array (`data_in`) of boolean values
+   * whose length is equal to the number of data fields specified in
+   * the config. It then encodes the data for servers A and B into a
+   * string.
+   *
+   * NOTE: The caller must free() the strings `for_server_a` and
+   * `for_server_b` to avoid memory leaks.
+   */
+  SECStatus PrioClient_encode(const_PrioConfig cfg,
+                              const bool* data_in,
+                              unsigned char** forServerA,
+                              unsigned int* aLen,
+                              unsigned char** forServerB,
+                              unsigned int* bLen);
+
+  /*
+   * Generate a new PRG seed using the NSS global randomness source.
+   * Use this routine to initialize the secret that the two Prio servers
+   * share.
+   */
+  SECStatus PrioPRGSeed_randomize(PrioPRGSeed* seed);
+
+  /*
+   * The PrioServer object holds the state of the Prio servers.
+   * Pass in the _same_ secret PRGSeed when initializing the two servers.
+   * The PRGSeed must remain secret to the two servers.
+   */
+  PrioServer PrioServer_new(const_PrioConfig cfg,
+                            PrioServerId serverIdx,
+                            PrivateKey serverPriv,
+                            const PrioPRGSeed serverSharedSecret);
+  void PrioServer_clear(PrioServer s);
+
+  /*
+   * After receiving a client packet, each of the servers generate
+   * a PrioVerifier object that they use to check whether the client's
+   * encoded packet is well formed.
+   */
+  PrioVerifier PrioVerifier_new(PrioServer s);
+  void PrioVerifier_clear(PrioVerifier v);
+
+  /*
+   * Read in encrypted data from the client, decrypt it, and prepare to check
+   * the request for validity.
+   */
+  SECStatus PrioVerifier_set_data(PrioVerifier v,
+                                  unsigned char* data,
+                                  unsigned int dataLen);
+
+  /*
+   * Generate the first packet that servers need to exchange to verify the
+   * client's submission. This should be sent over a TLS connection between the
+   * servers.
+   */
+  PrioPacketVerify1 PrioPacketVerify1_new(void);
+  void PrioPacketVerify1_clear(PrioPacketVerify1 p1);
+
+  SECStatus PrioPacketVerify1_set_data(PrioPacketVerify1 p1,
+                                       const_PrioVerifier v);
+
+  SECStatus PrioPacketVerify1_write(const_PrioPacketVerify1 p,
+                                    msgpack_packer* pk);
+  SECStatus PrioPacketVerify1_read(PrioPacketVerify1 p,
+                                   msgpack_unpacker* upk,
+                                   const_PrioConfig cfg);
+
+  /*
+   * Generate the second packet that the servers need to exchange to verify the
+   * client's submission. The routine takes as input the PrioPacketVerify1
+   * packets from both server A and server B.
+   *
+   * This should be sent over a TLS connection between the servers.
+   */
+  PrioPacketVerify2 PrioPacketVerify2_new(void);
+  void PrioPacketVerify2_clear(PrioPacketVerify2 p);
+
+  SECStatus PrioPacketVerify2_set_data(PrioPacketVerify2 p2,
+                                       const_PrioVerifier v,
+                                       const_PrioPacketVerify1 p1A,
+                                       const_PrioPacketVerify1 p1B);
+
+  SECStatus PrioPacketVerify2_write(const_PrioPacketVerify2 p,
+                                    msgpack_packer* pk);
+  SECStatus PrioPacketVerify2_read(PrioPacketVerify2 p,
+                                   msgpack_unpacker* upk,
+                                   const_PrioConfig cfg);
+
+  /*
+   * Use the PrioPacketVerify2s from both servers to check whether
+   * the client's submission is well formed.
+   */
+  SECStatus PrioVerifier_isValid(const_PrioVerifier v,
+                                 const_PrioPacketVerify2 pA,
+                                 const_PrioPacketVerify2 pB);
+
+  /*
+   * Each of the two servers calls this routine to aggregate the data
+   * submission from a client that is included in the PrioVerifier object.
+   *
+   * IMPORTANT: This routine does *not* check the validity of the client's
+   * data packet. The servers must execute the verification checks
+   * above before aggregating any client data.
+   */
+  SECStatus PrioServer_aggregate(PrioServer s, PrioVerifier v);
+
+  /*
+   * After the servers have aggregated data packets from "enough" clients
+   * (this determines the anonymity set size), each server runs this routine
+   * to get a share of the aggregate statistics.
+   */
+  PrioTotalShare PrioTotalShare_new(void);
+  void PrioTotalShare_clear(PrioTotalShare t);
+
+  SECStatus PrioTotalShare_set_data(PrioTotalShare t, const_PrioServer s);
+
+  SECStatus PrioTotalShare_write(const_PrioTotalShare t, msgpack_packer* pk);
+  SECStatus PrioTotalShare_read(PrioTotalShare t,
+                                msgpack_unpacker* upk,
+                                const_PrioConfig cfg);
+
+  /*
+   * Read the output data into an array of unsigned longs. You should
+   * be sure that each data value can fit into a single `unsigned long`
+   * and that the pointer `output` points to a buffer large enough to
+   * store one long per data field.
+   *
+   * This function returns failure if some final data value is too
+   * long to fit in an `unsigned long`.
+   */
+  SECStatus PrioTotalShare_final(const_PrioConfig cfg,
+                                 unsigned long long* output,
+                                 const_PrioTotalShare tA,
+                                 const_PrioTotalShare tB);
 
 #endif /* __PRIO_H__ */
 
--- a/pclient/main.c	2020-04-29 16:10:45.081830336 +0530
+++ b/pclient/main.c	2020-04-29 16:34:38.974314050 +0530
@@ -108,8 +108,8 @@
     //
     // Construct the client data packets.
     unsigned int aLen, bLen;
-    P_CHECKC(PrioClient_encode(cfg, data_items, &for_server_a, &aLen,
-                               &for_server_b, &bLen));
+    P_CHECKC(PrioClient_encode(
+      cfg, data_items, &for_server_a, &aLen, &for_server_b, &bLen));
 
     // The Prio servers A and B can come online later (e.g., at the end of
     // each day) to download the encrypted telemetry packets from the
--- a/prio/callbk.c	2020-04-29 16:10:45.081830336 +0530
+++ b/prio/callbk.c	2020-04-29 16:34:38.985923281 +0530
@@ -9,9 +9,11 @@
 #include "callbk.h"
 #include <nss.h>
 
-void nss_callbk_init(nss_struct *nss_pointer){
+void
+nss_callbk_init(nss_struct* nss_pointer)
+{
 
-    (nss_pointer)->nssinit = NSS_InitContext;
-    (nss_pointer)->nssisinit = NSS_IsInitialized;
-    (nss_pointer)->nssshutdown = NSS_ShutdownContext;
+  (nss_pointer)->nssinit = NSS_InitContext;
+  (nss_pointer)->nssisinit = NSS_IsInitialized;
+  (nss_pointer)->nssshutdown = NSS_ShutdownContext;
 }
\ No newline at end of file
--- a/prio/config.c	2020-04-29 16:10:45.081830336 +0530
+++ b/prio/config.c	2020-04-29 16:34:38.995175827 +0530
@@ -25,8 +25,11 @@
 }
 
 PrioConfig
-PrioConfig_new(int n_fields, PublicKey server_a, PublicKey server_b,
-               const unsigned char* batch_id, unsigned int batch_id_len)
+PrioConfig_new(int n_fields,
+               PublicKey server_a,
+               PublicKey server_b,
+               const unsigned char* batch_id,
+               unsigned int batch_id_len)
 {
   SECStatus rv = SECSuccess;
   PrioConfig cfg = malloc(sizeof(*cfg));
@@ -94,9 +97,17 @@
   return cfg->num_data_fields;
 }
 
-SECStatus Prio_init(nss_struct *nss_pointer) { return rand_init(nss_pointer); }
+SECStatus
+Prio_init(nss_struct* nss_pointer)
+{
+  return rand_init(nss_pointer);
+}
 
-void Prio_clear(nss_struct *nss_pointer) { rand_clear(nss_pointer); }
+void
+Prio_clear(nss_struct* nss_pointer)
+{
+  rand_clear(nss_pointer);
+}
 
 int
 PrioConfig_hPoints(const_PrioConfig cfg)
--- a/ptest/mutest.c	2020-04-29 16:10:45.081830336 +0530
+++ b/ptest/mutest.c	2020-04-29 16:34:39.005532132 +0530
@@ -81,14 +81,19 @@
 
   Prio_clear(&nss_pointer);
 
-  mu_print(MU_SUMMARY, "\n"
-                       "Tests done:\n"
-                       "\t%d test suite(s) passed, %d failed, %d skipped.\n"
-                       "\t%d test case(s) passed, %d failed.\n"
-                       "\t%d check(s) passed, %d failed.\n"
-                       "\n",
-           mutest_passed_suites, mutest_failed_suites, mutest_skipped_suites,
-           mutest_passed_cases, mutest_failed_cases, mutest_passed_checks,
+  mu_print(MU_SUMMARY,
+           "\n"
+           "Tests done:\n"
+           "\t%d test suite(s) passed, %d failed, %d skipped.\n"
+           "\t%d test case(s) passed, %d failed.\n"
+           "\t%d check(s) passed, %d failed.\n"
+           "\n",
+           mutest_passed_suites,
+           mutest_failed_suites,
+           mutest_skipped_suites,
+           mutest_passed_cases,
+           mutest_failed_cases,
+           mutest_passed_checks,
            mutest_failed_checks);
 
   return (mutest_failed_suites + mutest_skipped_suites) ? 1 : 0;

*** Aborting, please fix your commit(s) with 'git commit --amend' or 'git rebase -i <hash>'

shravanrn commented on prio/callbk.c:
```diff
@@ -0,0 +1,17 @@
+#include "callbk.h"
+#include <nss.h>
+
+void nss_callbk_init(nss_struct *nss_pointer){
```
Don't include references to NSS in the prio library. Move this code into pclient/main.c and ptest/mutest.c. It should be set only by users of libprio

shravanrn commented:
> Usually running git-clang-format and then uploading fixed my formatting, i tried to do some workaround and run script in ./script/clang-format.sh

Just run clang-format directly. 

1. Open terminal **in the libprio folder**
2. For each file you have modified, run `clang-format -i <relative_path_to_file>`

rhelmer commented:
> @Hyperion101010
> 
> > Usually running git-clang-format and then uploading fixed my formatting, i tried to do some workaround and run script in ./script/clang-format.sh
> 
> Just run clang-format directly.
> 
>     1. Open terminal **in the libprio folder**
> 
>     2. For each file you have modified, run `clang-format -i <relative_path_to_file>`

You probably need `clang-format -style=Mozilla -i` as well.

Commit:
[WIP]: Change prio for rlbox sandboxing.

    Formatting using mozila style and clang format.
    Removed callbk.c and added in driver code.

Hyperion101010 commented:
Used -style=Mozilla and clang-format lets see.

Commit:
[WIP]: Change prio for rlbox sandboxing #84

   Now without mozilla style formating.

Hyperion101010 commented:
Isuue looks like it will require changes in clang-format of libprio.
Robert opened one issue yesterday, I don't know about shell scripting 
much but hacking my local clang script file.
First lets update the clang file and then submit this patch.
What do you think @rhelmer  and @shravanrn

Commit:
[WIP]: Change prio for rlbox sandboxing

   Using clang-format 3.9 with style
   as Mozilla.

rhelmer commented:
> Isuue looks like it will require changes in clang-format of libprio.
> Robert opened one issue yesterday, I don't know about shell scripting
> much but hacking my local clang script file to see if i can run it for
> clang-format 9
> First lets update the clang file and then submit this patch.
> What do you think @rhelmer and @shravanrn

It looks like things are passing CI now, I think we can wait on issue #87 unless it's in your way. I just filed it so we don't have anyone else have to trip up on that older version of clang (sorry!)

Hyperion101010 commented:
@rhelmer I have clang-format 3.9 setup on mu pc now, also it is highly unlikely that someone else will start working on it (outside of usual contributors) 
I would like to get one final check from shravan and close this cause then I can start working on the Firefox end to add changes made by me and deian sir to sandbox prio. We can handle the update issue later with some free time. 
Let me know what you think, also no worries it is understandable that the repo is very rarely updated by any outside person and thus might bring in such deps issues.

shravanrn commented on pclient/main.c:
```diff
@@ -43,8 +55,12 @@ verify_full(void)
   bool* data_items = NULL;
 
+  // Initialize nss pointer for prio
+  static nss_struct nss_pointer;
```
Nit: no particular point in making this static if you are re-initializing it every time. I'd suggest just dropping the static, as this may be a race condition waiting to happen if the tests are parallelized.

shravanrn commented on ptest/mutest.c:
```diff
@@ -65,16 +68,28 @@ parse_args(__attribute__((unused)) int argc, char* argv[])
 {
 
-  Prio_init();
+  static nss_struct nss_pointer;
```
Nit: Same as other, drop the static.

shravanrn commented:
> @rhelmer I have clang-format 3.9 setup on my pc now, also it is highly unlikely that someone else will start working on it (outside of usual contributors)

I think Robert was talking about contributors to prio in general, not contributors to this patch :slightly_smiling_face: 

> I would like to get one final check from shravan 

Couple of minor things, but looks good overall. Once you fix those, I think it is ready to commit.

deian commented on include/callbk.h:
```diff
@@ -0,0 +1,35 @@
+  PRBool (*nssisinit)();
+
+  SECStatus (*nssshutdown)(NSSInitContext* context);
+} nss_struct;
```
Same. I'd give this a more meaningful name and also add a comment explaining why we need this.

deian commented on include/callbk.h:
```diff
@@ -0,0 +1,35 @@
+} nss_struct;
+
+/*
+nssinit = NSS_InitContext;
```
i'd axe this comment

deian reviewed (changes requested):
Good start. I think there are other bits in the prio directory that rely on NSS. Looking at [this](https://searchfox.org/mozilla-central/source/third_party/prio/update.sh) only the include and prio directories are getting pulled so we'd need to somehow expose the mpi stuff too

Hyperion101010 commented on include/mprio.h:
```diff
@@ -212,8 +214,8 @@ PrioVerifier PrioVerifier_new(PrioServer s);
 void PrioVerifier_clear(PrioVerifier v);
 
 /*
- * Read in encrypted data from the client, decrypt it, and prepare to check the
```
fixed, problems due to use of multiple format tools.

Hyperion101010 commented on .travis.yml:
```diff
@@ -4,18 +4,10 @@ os:
+
 before_install:
   - if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then
-      source ~/virtualenv/python2.7/bin/activate;
```
because its updated upstream.

Commit:
[WIP]: Change prio for rlbox sandboxing

   Changed the callbk file name to NssCtx,
   Added some comments and minor changes.

Commit:
[WIP]: Change prio for rlbox sandboxing

   Minor changes.

henrycg commented:
It looks like there hasn't been any activity on this PR for a while. If you're still interested in having it merged, feel free to submit a new one or reopen this one. Thanks!

deian commented:
Makes sense, sorry for not closing it earlier. I tabled it because of the rust implementation and i assume y'all don't have a need for this to be sandboxed for other reasons?

henrycg commented:
Yeah, as far as I can tell, the plan is to move to the Rust implementation. Even though this library will ship with Firefox for some time yet, it's probably not a high priority to sandbox.

deian commented:
Yeah my tabling it came from chatting with the Firefox folks. Thanks!
